httpx/_config.py

```diff
@@ -153,11 +154,15 @@ def _create_default_ssl_context(self) -> ssl.SSLContext:
         Creates the default SSLContext object that's used for both verified
         and unverified connections.
         """
-        context = ssl.SSLContext(ssl.PROTOCOL_TLS)
-        context.options |= ssl.OP_NO_SSLv2
-        context.options |= ssl.OP_NO_SSLv3
-        context.options |= ssl.OP_NO_TLSv1
-        context.options |= ssl.OP_NO_TLSv1_1
+        context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+        if sys.version_info >= (3, 10) or (
+            sys.version_info >= (3, 7) and ssl.OPENSSL_VERSION_INFO >= (1, 1, 0)
+        ):  # pragma: py-lt-37
+            context.minimum_version = ssl.TLSVersion.TLSv1_2
+        else:  # pragma: py-gte-37
+            context.options |= (
+                ssl.OP_NO_SSLv2 | ssl.OP_NO_SSLv3 | ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_1
+            )
```

Comment:
Can you talk me through the change here?

Why `or (sys.version_info >= (3, 7) and ssl.OPENSSL_VERSION_INFO >= (1, 1, 0))` and why `context.minimum_version = ssl.TLSVersion.TLSv1_2`?

Reply:
https://matrix.to/#/!IMqeNVkEzyjIEdEAul:gitter.im/$JX5LkGYztHop0XJc2OWeW67ik_pjZ2jtHXy1ZaxdWw8?via=matrix.org&via=gitter.im&via=chat.physics.ac

Reply:
I'll quote [your chat](https://gitter.im/encode/community?at=60c0864102f6ee5b91e1ee08) inline here...

> My understanding from chats with Tiran in #python is that the OP_NO flags are needed for openssl < (1,1,0)
> But py3.10 drops older versions of openssl and so can deprecate those old flags
> Also only py3.7+ exposes the new openssl 1,1,0 minimum_version interface
> The minimum version is 1.2 because that's what you get when you | those flags together, I think

Reply:
That feels awfully fuzzy.

I'd really rather we had a much clearer explanation and worked through rationale for changes like this, so that we're able to properly assess them, and so that there's  a clear working record for other folks digging into this.

From a quick glance over to URLLib3 https://github.com/urllib3/urllib3/commit/2ec06d1185f1642dfe8ff767a9ad3541bd0ddc39 it looks like using `PROTOCOL_TLS_CLIENT` instead might (or might not) be the right thing to do here.

I'd probably rather we were starting working though changes like this as "discussions" so that we can more fully thrash out *exactly* the behaviour and motivation.

Reply:
@tomchristie this PR does change the code to use PROTOCOL_TLS_CLIENT

Reply:
I think we'd do better to treat the one liner `ssl.PROTOCOL_TLS` -> `ssl.PROTOCOL_TLS_CLIENT` change as an isolated PR, without also bundling in the context branching, which looks to me like it needs a *separate* rationale/discussion.

Reply:
Okay, having had to resolve https://github.com/encode/httpx/pull/1687 I can see why you'd needed the context.options changes here.

I've approached things *slightly* differently there, using the `_compat.py` module for cases that need environment-based switches.